### PR TITLE
feat: migrate from rusoto to aws-sdk-rust

### LINKED ISSUE #51: Migrate from rusoto to AWS SDK for Rust
AWS SDK for Rust is now in developer preview.
https://aws.amazon.com/jp/about-aws/whats-new/2021/12/aws-sdk-rust-developer-preview/

For future GA, we start to consider migrating rusoto to AWS SDK for Rust.

### PR DESCRIPTION
**Pull request by ryota-sakamoto**
*Issue #, if available:*

close #51

*Description of changes:*

rusoto is in maintenance mode so we need to migrate from rusoto to aws-sdk-rust. We will address some actions after the PR is merged.

1. improve error message from dynein

We can show error message to use `into_service_error` but we need to consider what the informations are shown.

```
            error!("{}", e.into_service_error().meta());
```

2. separate config implementation from app.rs

### CODE REVIEW

**Comment by StoneDot** [comment]
This task is blocked by #143.

File: src/app.rs [hunk 1]
```diff
@@ -15,26 +15,28 @@
  */
 
 use ::serde::{Deserialize, Serialize};
+use aws_config::{meta::region::RegionProviderChain, BehaviorVersion, SdkConfig};
+use aws_sdk_dynamodb::types::{AttributeDefinition, TableDescription};
+use aws_types::region::Region;
```
Comment (StoneDot) [inline src/app.rs:20]: We can use `use aws_config::Region;`.

File: tests/util/mod.rs [hunk 2]
```diff
@@ -14,16 +14,17 @@
...
 
-use assert_cmd::prelude::*; // Add methods on commands
+use assert_cmd::prelude::*;
 use std::env;
 use std::process::Command; // Run programs
                            // use assert_cmd::cmd::Command; // Run programs - it seems to be equal to "use assert_cmd::prelude::* + use std::process::Command"
 
+use aws_config::SdkConfig;
+use aws_sdk_dynamodb::Client as DynamoDbSdkClient;
+use aws_types::region::Region;
```
Comment (StoneDot) [inline tests/util/mod.rs:24]: We can use `use aws_config::Region;`.

File: Cargo.toml [hunk 3]
```diff
@@ -30,6 +30,12 @@ name = "dy"
 path = "src/main.rs"
 
 [dependencies]
+aws-config       = "1.4.0"
+aws-sdk-dynamodb = "1.28.0"
+aws-sdk-ec2      = "1.42.0"
+aws-smithy-runtime-api = "1.6.0"
+aws-smithy-types = "1.1.9"
+aws-types        = "1.2.1"
```
Comment (StoneDot) [inline Cargo.toml:38]: We can remove this dependency because `aws_types::region::Region` is reexported as `aws_config::Region`.

**Review by StoneDot — COMMENTED** [review]
I am reviewing it. I left first quick comments.

File: src/control.rs [hunk 4]
```diff
@@ -178,15 +123,12 @@ pub async fn describe_table(cx: app::Context, target_table_to_desc: Option<Strin
...
-        new_context.effective_table_name(),
-    )
-    .await;
+    let desc: TableDescription =
+        describe_table_api(&new_context, new_context.effective_table_name()).await;
     debug!(
         "Retrieved table to describe is: '{}' table in '{}' region.",
         &new_context.effective_table_name(),
-        &new_context.effective_region().name()
+        &new_context.effective_region().as_ref()
```
Comment (StoneDot) [inline src/control.rs]: No need `&`.

File: tests/util/mod.rs [hunk 5]
```diff
@@ -338,15 +339,17 @@ async fn setup_container(port: i32) -> Result<(), Box<dyn std::error::Error>> {
...
 
     // Wait dynamodb-local
-    let health_check_url = format!("http://localhost:{}", port);
-    let ddb = DynamoDbClient::new(Region::Custom {
-        name: "local".to_owned(),
-        endpoint: health_check_url,
-    });
+    // https://docs.aws.amazon.com/sdk-for-rust/latest/dg/dynamodb-local.html
+    let config = aws_sdk_dynamodb::config::Builder::from(
+        &SdkConfig::builder().region(Region::new("local")).build(),
```
Comment (StoneDot) [inline tests/util/mod.rs:343]: Need to set `behavior_version`.
```suggestion
        &SdkConfig::builder()
            .region(Region::new("local"))
            .behavior_version(BehaviorVersion::v2024_03_28())
            .build(),
```

**Review by StoneDot — COMMENTED** [review]
Second iteration of the review.

File: src/app.rs [hunk 6]
```diff
@@ -344,9 +286,9 @@ impl Context {
 
         // otherwise, come down to "default region" of your environment.
         // e.g. region set via AWS CLI (check: $ aws configure get region), or environment variable `AWS_DEFAULT_REGION`.
-        //      ref: https://docs.rs/rusoto_signature/0.42.0/src/rusoto_signature/region.rs.html#282-290
         //      ref: https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-envvars.html
-        Region::default()
+        // TODO: fix
+        Region::from_static("us-east-1")
```
[the remark this answers is not in the document]
Comment (ryota-sakamoto) [reply src/app.rs]: fixed by https://github.com/awslabs/dynein/pull/126/commits/704cb0c7841743f8fc394af6a24caca8c28a33d8

File: src/app.rs [hunk 7]
```diff
@@ -908,26 +846,22 @@ mod tests {
...
-            .with_jitter()
-            .with_factor(2.0)
-            .with_min_delay(Duration::from_secs(1))
-            .with_max_delay(Duration::from_secs(100))
-            .with_max_times(19);
+        let actual = RetryConfig::try_from(config2).unwrap();
+        let expected = RetryConfig::standard()
+            .with_initial_backoff(Duration::from_secs(1))
+            .with_max_backoff(Duration::from_secs(100))
+            .with_max_attempts(19);
```
Comment (StoneDot) [inline src/app.rs:864]: need modification to `20`.

File: src/app.rs [hunk 8]
```diff
@@ -908,26 +846,22 @@ mod tests {
...
-        let actual = ExponentialBuilder::try_from(config1).unwrap();
-        let expected = ExponentialBuilder::default()
-            .with_min_delay(Duration::from_secs(1))
-            .with_jitter()
-            .with_factor(2.0)
-            .with_max_times(9);
+        let actual = RetryConfig::try_from(config1).unwrap();
+        let expected = RetryConfig::standard()
+            .with_initial_backoff(Duration::from_secs(1))
+            .with_max_attempts(9);
```
Comment (StoneDot) [inline src/app.rs:852]: need modification to `10`.

File: src/ddb/mod.rs
Comment (StoneDot) [inline src/ddb/mod.rs:1]: Could you place a copyright notice?

File: src/key.rs
Comment (StoneDot) [inline src/key.rs:1]: This can be a submodule of `ddb`.

**Review by StoneDot — COMMENTED** [review]
I have finalized my review. Could you please review the comments I left? Thank you for your hard work on this pull request.

**Comment by ryota-sakamoto** [comment]
@StoneDot 

Thank you for review. I have addressed all of the points you mentioned.

**Review by StoneDot — APPROVED** [review]
I will approve this. I appreciate your contribution.

**Pull request merged**